accelerator/core/apis.c

```diff
@@ -426,101 +426,99 @@ status_t api_send_mam_message(const ta_config_t* const info, const iota_config_t
-  if (ret) {
-    ta_log_error("%d\n", ret);
-    goto done;
-  }
-  ret = send_mam_res_set_msg_id(res, msg_id);
-  if (ret) {
-    ta_log_error("%d\n", ret);
-    goto done;
-  }
-  ret = send_mam_res_set_bundle_hash(res, transaction_bundle((iota_transaction_t*)utarray_front(bundle)));
-  if (ret) {
-    ta_log_error("%d\n", ret);
+    ta_log_error("%s\n", ta_error_to_string(ret));
     goto done;
   }
-
-  bundle_transactions_free(&bundle);
-  bundle_transactions_new(&bundle);
-
-  // Sending bundle
-  if (ch_remain_sk == 1 || ep_remain_sk == 1) {
-    // Send announcement for the next endpoint or channel
-    ret = ta_mam_announce_next_mss_private_key_to_bundle(&mam, data->ch_mss_depth, data->ep_mss_depth, chid,
-                                                         &ch_remain_sk, &ep_remain_sk, psks, ntru_pks, chid1, &bundle);
-    if (ret) {
-      ta_log_error("%d\n", ret);
+  mam_mss_key_status_t key_status;
+  while (!msg_sent) {
+    bundle_transactions_renew(&bundle);
+
+    // Create epid merkle tree and find the smallest unused secret key.
+    // Write both Header and Pakcet into one single bundle.
+    ret = ta_mam_written_msg_to_bundle(service, &mam, data->ch_mss_depth, chid, psks, ntru_pks, data->message, &bundle,
+                                       msg_id, &key_status);
+    if (ret == SC_OK) {
+      msg_sent = true;
+    } else if (ret == SC_MAM_ALL_MSS_KEYS_USED) {
+      ta_log_debug("%s\n", ta_error_to_string(ret));
+      goto announce;
+    } else {
+      ta_log_error("%s\n", ta_error_to_string(ret));
       goto done;
     }
 
+    // Sending bundle
     lock_handle_lock(&cjson_lock);
-    if (ta_send_bundle(info, iconf, service, bundle) != SC_OK) {
+    ret = ta_send_bundle(info, iconf, service, bundle);
+    if (ret != SC_OK) {
       lock_handle_unlock(&cjson_lock);
-      ret = SC_MAM_FAILED_RESPONSE;
-      ta_log_error("%s\n", "SC_MAM_FAILED_RESPONSE");
+      ta_log_error("%s\n", ta_error_to_string(ret));
       goto done;
     }
     lock_handle_unlock(&cjson_lock);
 
-    ret =
-        send_mam_res_set_announcement_bundle_hash(res, transaction_bundle((iota_transaction_t*)utarray_front(bundle)));
+    ret = send_mam_res_set_channel_id(res, chid);
     if (ret) {
-      ta_log_error("%d\n", ret);
+      ta_log_error("%s\n", ta_error_to_string(ret));
       goto done;
     }
 
-    ret = send_mam_res_set_chid1(res, chid1);
+    ret = send_mam_res_set_msg_id(res, msg_id);
     if (ret) {
-      ta_log_error("%d\n", ret);
+      ta_log_error("%s\n", ta_error_to_string(ret));
       goto done;
     }
+
+    tryte_t bundle_hash[NUM_TRYTES_BUNDLE] = {};
+    flex_trits_to_trytes(bundle_hash, NUM_TRYTES_BUNDLE, transaction_bundle((iota_transaction_t*)utarray_front(bundle)),
+                         NUM_TRITS_BUNDLE, NUM_TRITS_BUNDLE);
+    ret = send_mam_res_set_bundle_hash(res, bundle_hash);
+    if (ret) {
+      ta_log_error("%s\n", ta_error_to_string(ret));
+      goto done;
+    }
+
+  announce:
```

Comment:
I am not sure if `announce` makes sense of not.

Reply:
How about renaming it as `mam_announce`?

Reply:
It sounds better.